src/mixins/java/org/spongepowered/common/mixin/core/server/network/ServerGamePacketListenerImplMixin.java

```diff
@@ -288,42 +295,37 @@
+        final Vector3d originalToPosition = new Vector3d(packetIn.getX(this.player.getX()),
                 packetIn.getY(this.player.getY()), packetIn.getZ(this.player.getZ()));
-        Vector3d toRotation = new Vector3d(packetIn.getYRot(this.player.yRot),
+        final Vector3d originalToRotation = new Vector3d(packetIn.getYRot(this.player.yRot),
                 packetIn.getXRot(this.player.xRot), 0);
 
-        final boolean significantRotation = fromRotation.distanceSquared(toRotation) > (.15f * .15f);
-
-        final Vector3d originalToPosition = toPosition;
-        boolean cancelMovement = false;
-        boolean cancelRotation = false;
-        // Call move & rotate event as needed...
+        // common checks and throws are done here.
+        final @Nullable Vector3d toPosition;
         if (fireMoveEvent) {
-            final MoveEntityEvent event = SpongeEventFactory.createMoveEntityEvent(PhaseTracker.getCauseStackManager().currentCause(), (ServerPlayer) this.player, fromPosition,
-                    toPosition, toPosition);
-            if (SpongeCommon.post(event)) {
-                cancelMovement = true;
-            } else {
-                toPosition = event.destinationPosition();
-            }
+            toPosition = SpongeCommonEventFactory.callMoveEvent(
+                    player,
+                    fromPosition,
+                    originalToPosition);
+        } else {
+            toPosition = null;
         }
 
-        if (significantRotation && fireRotationEvent) {
-            final RotateEntityEvent event = SpongeEventFactory.createRotateEntityEvent(PhaseTracker.getCauseStackManager().currentCause(), (ServerPlayer) this.player, fromRotation,
-                    toRotation);
-            if (SpongeCommon.post(event)) {
-                cancelRotation = true;
-                toRotation = fromRotation;
-            } else {
-                toRotation = event.toRotation();
-            }
+        // Rotation result
+        final Vector3d toRotation;
+        if (fireRotationEvent) {
+            toRotation = SpongeCommonEventFactory.callRotateEvent(
+                    player,
+                    fromRotation,
+                    originalToRotation);
+        } else {
+            toRotation = null;
         }
 
         // At this point, we cancel out and let the "confirmed teleport" code run through to update the
         // player position and update the player's relation in the chunk manager.
-        if (cancelMovement) {
-            if (fromPosition.distanceSquared(toPosition) > 0) {
+        if (toPosition == null) {
+            if (fromPosition.distanceSquared(originalToPosition) > 0) {
                 // Set the location, as if the player was teleporting
                 this.awaitingTeleportTime = this.tickCount;
                 this.shadow$teleport(fromPosition.x(), fromPosition.y(), fromPosition.z(), (float) toRotation.x(), (float) toRotation.y());
```

Comment:
If `toRotation` and `toPosition` are both null, then this `toRotation` call here throws an NPE. Should there be something around line `321` that says "if `toRotation` is null, make it `fromLocation`"?

Reply:
Yup, good spot.